[backport] docs: early October improvements and fixes

icyJoseph opened the pull request:
Additional group of docs improvements.

icyJoseph committed:
Docs/getting started improv (#82741)

Part of:
https://linear.app/vercel/issue/DOC-4809/layout-pages-and-linking-and-navigating-improvements

Iterating a bunch of improvements to getting-started

geraldochristiano committed:
Fixed broken link in backend-for-frontend.mdx (#83310)

Broken link to API Routes in Pages Router

Co-authored-by: JJ Kasper <jj@jjsweb.site>

icyJoseph committed:
docs: We have dropped the CORS example linked (#84065)

We have dropped the example linked in this page. I'll re-evaluate later
if the section is still enough for a dev to understand how to set CORS
related headers.

Milancen123 committed:
docs: add pageExtensions note to Middleware and clarify default filename (#84045)

- Added a Note to the Middleware docs explaining that when
`pageExtensions` are customized (e.g. `.page.ts`), the Middleware file
should be named `middleware.page.ts` or `middleware.page.js`.
- Updated wording so `middleware.ts/js` is described as the default
filename, not a strict requirement.

<!-- Thanks for opening a PR! Your contribution is much appreciated.
To make sure your PR is handled as smoothly as possible we request that
you follow the checklist sections below.
Choose the right checklist for the change(s) that you're making:

## For Contributors

### Improving Documentation

- Run `pnpm prettier-fix` to fix formatting issues before opening the
PR.
- Read the Docs Contribution Guide to ensure your contribution follows
the docs guidelines:
https://nextjs.org/docs/community/contribution-guide

### Fixing a bug

- Related issues linked using `fixes #number`
- Tests added. See:
https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs
- Errors have a helpful link attached, see
https://github.com/vercel/next.js/blob/canary/contributing.md

### Adding a feature

- Implements an existing feature request or RFC. Make sure the feature
request has been accepted for implementation before opening a PR. (A
discussion must be opened, see
https://github.com/vercel/next.js/discussions/new?category=ideas)
- Related issues/discussions are linked using `fixes #number`
- e2e tests added
(https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs)
- Documentation added
- Telemetry added. In case of a feature if it's used or not.
- Errors have a helpful link attached, see
https://github.com/vercel/next.js/blob/canary/contributing.md


## For Maintainers

- Minimal description (aim for explaining to someone not on the team to
understand the PR)
- When linking to a Slack thread, you might want to share details of the
conclusion
- Link both the Linear (Fixes NEXT-xxx) and the GitHub issues
- Add review comments if necessary to explain to the reviewer the logic
behind a change

### What?

-Added a callout note to middleware.mdx stating that if pageExtensions
is customized (e.g., .page.ts), the Middleware file should also match
(middleware.page.ts/js).
-Updated the language from “must” → “default is” for filename clarity.


### Why?

-Developers customizing pageExtensions were misled by the existing docs,
which implied middleware.ts/js was always required.
-This PR prevents confusion when Middleware isn’t recognized after a
pageExtensions change.


### How?
Edited docs/app/api-reference/file-conventions/middleware.mdx.
Inserted a > **Note** block right after the paragraph explaining where
to create the middleware file.


Closes NEXT-
Fixes #84008

-->

---------

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

martinfrancois committed:
Fix TypeScript type error in example code (#84041)

Without the `| null`, the following error appears during compilation:
```
TS2322: Type RefObject<HTMLDivElement | null> is not assignable to type RefObject<HTMLDivElement>
Type HTMLDivElement | null is not assignable to type HTMLDivElement
Type null is not assignable to type HTMLDivElement
```

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

liketiger committed:
docs: improve dynamic routes example in project structure page (#84086)

## Documentation Update

This PR improves the dynamic routes example in the
`project-structure.mdx` page.

### Changes
- Updated catch-all segments example:
  - Before: `/shop/clothing/shoes`
  - After:  `/shop/clothing`
- Adjusted follow-up examples for consistency:
  - `/shop/a`, `/shop/a/b` → `/shop/clothing`, `/shop/clothing/shirts`

### Why
- Provides a clearer and more intuitive demonstration of how catch-all
segments work.
- Keeps the example consistent with other dynamic routes documentation.

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

icyJoseph committed:
docs: onRequestError request header type (#83988)

Aligning with the actual type
https://github.com/vercel/next.js/blob/canary/packages/next/src/server/instrumentation/types.ts#L18

icyJoseph committed:
docs: nav_title for long unbroken words (#84233)

Let's use a different title for these, otherwise we have a very long
composite word.

ijjk commented:
## Failing test suites <!--47380343150-->
Commit: be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2 | [About building and testing Next.js](https://github.com/vercel/next.js/blob/canary/contributing.md)

`pnpm test-start-turbo test/e2e/app-dir/back-forward-cache/back-forward-cache.test.ts` (turbopack)
- back/forward cache > React state is preserved when navigating with back/forward buttons ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22back%2Fforward+cache+React+state+is+preserved+when+navigating+with+back%2Fforward+buttons%22+%40test.type%3A%22turbopack%22+%40test.status%3A%22fail%22))
- back/forward cache > React state is preserved when navigating back/forward with links ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22back%2Fforward+cache+React+state+is+preserved+when+navigating+back%2Fforward+with+links%22+%40test.type%3A%22turbopack%22+%40test.status%3A%22fail%22))
- back/forward cache > React state is preserved when navigating back to a page with different search params than before  ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22back%2Fforward+cache+React+state+is+preserved+when+navigating+back+to+a+page+with+different+search+params+than+before+%22+%40test.type%3A%22turbopack%22+%40test.status%3A%22fail%22))
- back/forward cache > bfcache only preserves up to N entries ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22back%2Fforward+cache+bfcache+only+preserves+up+to+N+entries%22+%40test.type%3A%22turbopack%22+%40test.status%3A%22fail%22))
- back/forward cache > navigate back and forth repeatedly between the same pages without evicting ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22back%2Fforward+cache+navigate+back+and+forth+repeatedly+between+the+same+pages+without+evicting%22+%40test.type%3A%22turbopack%22+%40test.status%3A%22fail%22))

<details>
<summary>Expand output</summary>

  ● back/forward cache › React state is preserved when navigating with back/forward buttons

    next build failed with code/signal 1

      107 |             if (code || signal)
      108 |               reject(
    > 109 |                 new Error(
          |                 ^
      110 |                   `next build failed with code/signal ${code || signal}`
      111 |                 )
      112 |               )

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:109:17)

  ● back/forward cache › React state is preserved when navigating back/forward with links

    next build failed with code/signal 1

      107 |             if (code || signal)
      108 |               reject(
    > 109 |                 new Error(
          |                 ^
      110 |                   `next build failed with code/signal ${code || signal}`
      111 |                 )
      112 |               )

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:109:17)

  ● back/forward cache › React state is preserved when navigating back to a page with different search params than before 

    next build failed with code/signal 1

      107 |             if (code || signal)
      108 |               reject(
    > 109 |                 new Error(
          |                 ^
      110 |                   `next build failed with code/signal ${code || signal}`
      111 |                 )
      112 |               )

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:109:17)

  ● back/forward cache › bfcache only preserves up to N entries

    next build failed with code/signal 1

      107 |             if (code || signal)
      108 |               reject(
    > 109 |                 new Error(
          |                 ^
      110 |                   `next build failed with code/signal ${code || signal}`
      111 |                 )
      112 |               )

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:109:17)

  ● back/forward cache › navigate back and forth repeatedly between the same pages without evicting

    next build failed with code/signal 1

      107 |             if (code || signal)
      108 |               reject(
    > 109 |                 new Error(
          |                 ^
      110 |                   `next build failed with code/signal ${code || signal}`
      111 |                 )
      112 |               )

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:109:17)

</details>

`pnpm test test/integration/custom-page-extension/test/index.test.js`
- Custom page extension > development mode > should work with normal page ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22Custom+page+extension+development+mode+should+work+with+normal+page%22+%40test.type%3A%22nextjs%22+%40test.status%3A%22fail%22))
- Custom page extension > development mode > should work dynamic page ([DD](https://app.datadoghq.com/ci/test/runs?query=%40git.repository.id%3A%22github.com%2Fvercel%2Fnext.js%22+%40git.commit.head_sha%3A%22be6f71d4f9885e47b7fc2fc7fd957a69eefb1ec2%22+%40test.name%3A%22Custom+page+extension+development+mode+should+work+dynamic+page%22+%40test.type%3A%22nextjs%22+%40test.status%3A%22fail%22))

<details>
<summary>Expand output</summary>

  ● Custom page extension › development mode › should work with normal page

    FetchError: request to http://localhost:38287/blog failed, reason: socket hang up

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7_encoding@0.1.13/node_modules/node-fetch/lib/index.js:1491:11)

  ● Custom page extension › development mode › should work dynamic page

    FetchError: request to http://localhost:38287/blog/nextjs failed, reason: connect ECONNREFUSED 127.0.0.1:38287

      at ClientRequest.<anonymous> (../node_modules/.pnpm/node-fetch@2.6.7_encoding@0.1.13/node_modules/node-fetch/lib/index.js:1491:11)

</details>

icyJoseph committed:
Revert "docs: nav_title for long unbroken words (#84233)"

This reverts commit 895f0f5b6a2a11e2107568e0547623a5fd1fc0b1.

Anshuman71 committed:
docs: remove repeated to (#84325)

### What?

Removed an extra "to"

### Why?

Semantics error/not required

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

icyJoseph committed:
docs: Type helpers don't need to be imported (#84265)

Let's be more clear about the usage of these types.

joao4xz committed:
fix: Missing link in Lazy Loading docs (#84473)

Add missing space before "Importing Client Components" in Lazy Loading
/app docs.

This fixes the missing link in "On this page" sidebar:

<img width="1092" height="446" alt="image"
src="https://github.com/user-attachments/assets/0981467d-ad61-494d-80b2-c0a0e323dd0b"
/>

In the current state, the first link is "Skipping SSR" (second link)
instead of "Importing Client Components"

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

icyJoseph committed:
docs: simplify streaming metadata (#84253)

Co-authored-by: vercel[bot] <35613825+vercel[bot]@users.noreply.github.com>

ijjk committed:
Add middleware to routes order doc (#84618)

This adds `middleware` in the routes order to clarify where it comes in
priority.

karlhorky committed:
Add note about $ACTION_ properties, remove reference to .entries() (#84459)

<!-- Thanks for opening a PR! Your contribution is much appreciated.
To make sure your PR is handled as smoothly as possible we request that
you follow the checklist sections below.
Choose the right checklist for the change(s) that you're making:

## For Contributors

### Improving Documentation

- Run `pnpm prettier-fix` to fix formatting issues before opening the
PR.
- Read the Docs Contribution Guide to ensure your contribution follows
the docs guidelines:
https://nextjs.org/docs/community/contribution-guide

### Fixing a bug

- Related issues linked using `fixes #number`
- Tests added. See:
https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs
- Errors have a helpful link attached, see
https://github.com/vercel/next.js/blob/canary/contributing.md

### Adding a feature

- Implements an existing feature request or RFC. Make sure the feature
request has been accepted for implementation before opening a PR. (A
discussion must be opened, see
https://github.com/vercel/next.js/discussions/new?category=ideas)
- Related issues/discussions are linked using `fixes #number`
- e2e tests added
(https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs)
- Documentation added
- Telemetry added. In case of a feature if it's used or not.
- Errors have a helpful link attached, see
https://github.com/vercel/next.js/blob/canary/contributing.md


## For Maintainers

- Minimal description (aim for explaining to someone not on the team to
understand the PR)
- When linking to a Slack thread, you might want to share details of the
conclusion
- Link both the Linear (Fixes NEXT-xxx) and the GitHub issues
- Add review comments if necessary to explain to the reviewer the logic
behind a change

### What?

### Why?

### How?

Closes NEXT-
Fixes #

-->

### What?

Make a note about the `$ACTION_` properties in `formData`

### Why?

Starting with next@14.1.3, `$ACTION_` properties have appeared in the
`formData`, which cause validation errors when used with a strict
validation with a library such as Zod or Yup:

```tsx
// Form Client Component
export default function Form() {
  const [state, formAction] = useFormState(action, {
    type: 'initial',
  });

  return (
    <form action={action}>
      {/* ... */}
    </form>
  );
}

// Server Action
export async function action(
  prevState: ServerActionReturnValue,
  formData: FormData,
): Promise<ServerActionReturnValue> {
  const validationResult = await schema.validate(
    Object.fromEntries(formData.entries()),
  );
```

The errors from Yup:

Errors:

- this field has unspecified keys: $ACTION_REF_1, $ACTION_1:0,
$ACTION_1:1, $ACTION_KEY
The data in the formData.entries()

```js
{
  '$ACTION_REF_1': '',
  '$ACTION_1:0': '{"id":"59f475b...","bound":"$@1"}',
  '$ACTION_1:1': '[{"type":"initial"}]',
  '$ACTION_KEY': 'k187677481',
  // Real form data starts here:
  year: '2024',
```

### How?

1. Document additional properties with keys starting with the prefix
`$ACTION_`
2. Remove the obsolete mention of `.entries()`, since that is not being
used in the code example anymore (previously, it was `const rawFormData
= Object.fromEntries(formData.entries())`)

### History

- The `$ACTION_` property change was added in
https://github.com/vercel/next.js/pull/63048
- The `docs/01-app/02-guides/forms.mdx` file was added in PR
https://github.com/vercel/next.js/pull/79288 , without the `$ACTION_`
property change

cc @delbaoliveira

Co-authored-by: Joseph <joseph.chamochumbi@vercel.com>

icyJoseph committed:
Merge branch 'next-15-5' into next-15-5-docs-backport-1

icyJoseph committed:
fix: docs formatting w/ whitespace
